transports/tls/tls.go

```diff
@@ -1,30 +1,162 @@
+// +build go1.8
+
 package tls
 
 import (
 	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"log"
 	"net"
+	"os"
+	"strings"
 
 	"github.com/gliderlabs/logspout/adapters/raw"
 	"github.com/gliderlabs/logspout/router"
 )
 
+const (
+	// constants used to identify environment variable names
+	envDisableSystemRoots = "LOGSPOUT_TLS_DISABLE_SYSTEM_ROOTS"
+	envCaCerts            = "LOGSPOUT_TLS_CA_CERTS"
+	envClientCert         = "LOGSPOUT_TLS_CLIENT_CERT"
+	envClientKey          = "LOGSPOUT_TLS_CLIENT_KEY"
+	envTLSHardening       = "LOGSPOUT_TLS_HARDENING"
+)
+
+var (
+	// package wide cache of TLS config
+	clientTLSConfig *tls.Config
+	// PCI compliance as of Jun 30, 2018: anything under TLS 1.1 must be disabled
+	// we bump this up to TLS 1.2 so we can support best possible ciphers
+	hardenedMinVersion = uint16(tls.VersionTLS12)
+	// allowed ciphers when in hardened mode
+	// disable CBC suites (Lucky13 attack) this means TLS 1.1 can't work (no GCM)
+	// only use perfect forward secrecy ciphers
+	hardenedCiphers = []uint16{
+		tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+		tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+		tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+		tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+		// these ciphers require go 1.8+
+		tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
+		tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
+	}
+	// EC curve preference when in hardened mode
+	// curve reference: http://safecurves.cr.yp.to/
+	hardenedCurvePreferences = []tls.CurveID{
+		// this curve is a non-NIST curve with no NSA influence. Prefer this over all others!
+		// this curve required go 1.8+
+		tls.X25519,
+		// These curves are provided by NIST; prefer in descending order
+		tls.CurveP521,
+		tls.CurveP384,
+		tls.CurveP256,
+	}
+)
+
+type tlsTransport int
+
 func init() {
 	router.AdapterTransports.Register(new(tlsTransport), "tls")
 	// convenience adapters around raw adapter
 	router.AdapterFactories.Register(rawTLSAdapter, "tls")
+
+	// we should load our TLS configuration only once
+	// since it is not expected to change during runtime
+	var err error
+	clientTLSConfig, err = createTLSConfig()
+
+	// without a valid/desired TLS config, we should exit
+	if err != nil {
+		log.Fatalf("error with TLSConfig: %s", err)
+	}
 }
 
 func rawTLSAdapter(route *router.Route) (router.LogAdapter, error) {
 	route.Adapter = "raw+tls"
 	return raw.NewRawAdapter(route)
 }
 
-type tlsTransport int
-
 func (t *tlsTransport) Dial(addr string, options map[string]string) (net.Conn, error) {
-	conn, err := tls.Dial("tcp", addr, nil)
+	// at this point, if our trust store is empty, there is no point of continuing
+	// since it would be impossible to successfully validate any x509 server certificates
+	if len(clientTLSConfig.RootCAs.Subjects()) < 1 {
+		return nil, fmt.Errorf("FATAL: TLS CA trust store is empty! Can not trust any TLS endpoints: tls://%s", addr)
```

Comment:
i generally prefer named return vars such that this would become:
```
    err = fmt.Errorf("FATAL: TLS CA trust store is empty! Can not trust any TLS endpoints: tls://%s", addr)
    return
```

Reply:
OK no problem. Chaning to:

```
if len(clientTLSConfig.RootCAs.Subjects()) < 1 {
	err := fmt.Errorf("FATAL: TLS CA trust store is empty! Can not trust any TLS endpoints: tls://%s", addr)
	return nil, err
}
```

Reply:
ah i actually meant to say perhaps we should use the following style throughout this file (and subsequently the project)
```
func (t *tlsTransport) Dial(addr string, options map[string]string) (conn net.Conn, err error) {
  if len(clientTLSConfig.RootCAs.Subjects()) < 1 {
    err = fmt.Errorf("FATAL: TLS CA trust store is empty! Can not trust any TLS endpoints: tls://%s", addr)
    return
  }
  // attempt to establish the TLS connection
  if conn, err = tls.Dial("tcp", addr, clientTLSConfig); err != nil{
    return
  }
  return
}
```
note the named return variables in the function definition and the use of `return` without args. thoughts?

Reply:
@michaelshobbs that's how I prefer it as well. The reason I didn't name my return values was because I noticed that the rest of the project does not and I was trying to be consistent. I will change it to that style now.